python/ray/tune/syncer.py

```diff
@@ -0,0 +1,253 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import distutils
+import logging
+import os
+import subprocess
+import tempfile
+import time
+import types
+
+try:  # py3
+    from shlex import quote
+except ImportError:  # py2
+    from pipes import quote
+
+from ray.tune.sample import function as tune_function
+from ray.tune.error import TuneError
+from ray.tune.log_sync import log_sync_template, NodeSyncMixin
+
+logger = logging.getLogger(__name__)
+
+S3_PREFIX = "s3://"
+GS_PREFIX = "gs://"
+ALLOWED_REMOTE_PREFIXES = (S3_PREFIX, GS_PREFIX)
+
+_syncers = {}
+
+
+def validate_sync_string(sync_string):
+    if "{source}" not in sync_string:
+        raise ValueError("Sync template missing '{source}'.")
+    if "{target}" not in sync_string:
+        raise ValueError("Sync template missing '{target}'.")
+
+
+def wait_for_sync():
+    for syncer in _syncers.values():
+        syncer.wait()
+
+
+class BaseSyncer(object):
+    def __init__(self, local_dir, remote_dir, sync_function=None):
+        """
+        Arguments:
+            local_dir (str): Directory to sync. Uniquely identifies the syncer.
+            remote_dir (str): Remote directory to sync with.
+            sync_function (func): Function for syncing the local_dir to
+                remote_dir. Defaults to a Noop.
+        """
+        if local_dir:
```

Comment:
```suggestion
        self._local_dir = (
            self._local_dir = os.path.join(local_dir, "")
            if local_dir
            else local_dir)
```

Reply:
Should this explicitly check `if local_dir is None`? Seems like with `local_dir == ""`, this could end up being `"/"`, which might be problematic.

Reply:
"" evaluates to false though